File: app01_java_nft/backend/src/main/java/com/devpowers/nftcerts/certificate/Artwork.java
package com.devpowers.nftcerts.certificate;

import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.EnumType;
import jakarta.persistence.Enumerated;
import jakarta.persistence.Id;
import jakarta.persistence.Index;
import jakarta.persistence.Table;

import java.time.Instant;
import java.util.UUID;

@Entity
@Table(name = "artwork", indexes = @Index(name = "ux_artwork_sha256", columnList = "sha256Hash", unique = true))
public class Artwork {

    @Id
    private UUID id;

    @Column(nullable = false)
    private UUID fileId;

    @Column(nullable = false)
    private String originalFilename;

    @Column(nullable = false, unique = true, length = 64)
    private String sha256Hash;

    private String title;

    @Column(length = 4000)
    private String description;

    private String medium;

    private Integer yearCreated;

    @Column(nullable = false)
    private Integer royaltyPercentageBps;

    @Column(nullable = false)
    private String artistWalletAddress;

    private String artistDid;

    private String imageIpfsUri;

    private String metadataIpfsUri;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private ArtworkStatus status;

    @Column(nullable = false)
    private Instant createdAt;

    protected Artwork() {
        // JPA
    }

    public Artwork(UUID id, UUID fileId, String originalFilename, String sha256Hash, String title,
                   String description, String medium, Integer yearCreated, Integer royaltyPercentageBps,
                   String artistWalletAddress, String artistDid, ArtworkStatus status, Instant createdAt) {
        this.id = id;
        this.fileId = fileId;
        this.originalFilename = originalFilename;
        this.sha256Hash = sha256Hash;
        this.title = title;
        this.description = description;
        this.medium = medium;
        this.yearCreated = yearCreated;
        this.royaltyPercentageBps = royaltyPercentageBps;
        this.artistWalletAddress = artistWalletAddress;
        this.artistDid = artistDid;
        this.status = status;
        this.createdAt = createdAt;
    }

    public UUID getId() {
        return id;
    }

    public UUID getFileId() {
        return fileId;
    }

    public String getOriginalFilename() {
        return originalFilename;
    }

    public String getSha256Hash() {
        return sha256Hash;
    }

    public String getTitle() {
        return title;
    }

    public String getDescription() {
        return description;
    }

    public String getMedium() {
        return medium;
    }

    public Integer getYearCreated() {
        return yearCreated;
    }

    public Integer getRoyaltyPercentageBps() {
        return royaltyPercentageBps;
    }

    public String getArtistWalletAddress() {
        return artistWalletAddress;
    }

    public String getArtistDid() {
        return artistDid;
    }

    public String getImageIpfsUri() {
        return imageIpfsUri;
    }

    public void setImageIpfsUri(String imageIpfsUri) {
        this.imageIpfsUri = imageIpfsUri;
    }

    public String getMetadataIpfsUri() {
        return metadataIpfsUri;
    }

    public void setMetadataIpfsUri(String metadataIpfsUri) {
        this.metadataIpfsUri = metadataIpfsUri;
    }

    public ArtworkStatus getStatus() {
        return status;
    }

    public void setStatus(ArtworkStatus status) {
        this.status = status;
    }

    public Instant getCreatedAt() {
        return createdAt;
    }
}

File: app03_csharp_nft/backend/NftCerts/Certificates/Artwork.cs
using System.ComponentModel.DataAnnotations;
using Microsoft.EntityFrameworkCore;

namespace NftCerts.Certificates;

public enum ArtworkStatus
{
    UPLOADED,
    PINNED,
    MINTED,
}

// Mirrors app01's Artwork JPA entity: an uploaded photograph after hashing and watermarking,
// before (or after) it has been pinned to IPFS and minted.
[Index(nameof(Sha256Hash), IsUnique = true)]
public class Artwork
{
    [Key]
    public string Id { get; set; } = "";          // UUID
    public string FileId { get; set; } = "";       // UUID
    public string OriginalFilename { get; set; } = "";
    public string Sha256Hash { get; set; } = "";   // hex, 64 chars, unique
    public string Title { get; set; } = "";
    public string Description { get; set; } = "";
    public string Medium { get; set; } = "";
    public int? YearCreated { get; set; }
    public int RoyaltyPercentageBps { get; set; }
    public string ArtistWalletAddress { get; set; } = "";
    public string ArtistDid { get; set; } = "";
    public string ImageIpfsUri { get; set; } = "";
    public string MetadataIpfsUri { get; set; } = "";
    public ArtworkStatus Status { get; set; } = ArtworkStatus.UPLOADED;
    public string CreatedAt { get; set; } = "";    // ISO-8601 instant
}
